server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java

```diff
@@ -299,7 +317,8 @@ private static Set<String> listOfNewFileNames(IndexCommit previous, IndexCommit
      * Checks whether the deletion policy is holding on to acquired index commits
      */
     synchronized boolean hasAcquiredIndexCommits() {
-        return acquiredIndexCommits.isEmpty() == false;
+        return acquiredIndexCommits.isEmpty() == false
+            && acquiredIndexCommits.keySet().stream().anyMatch(Predicate.not(indexCommitsAcquiredByCommitsListener::contains));
```

Comment:
I'd find it slightly more logical to not add internally acquired commits into `acquiredIndexCommits` (and perhaps rename that to `externallyAcquiredIndexCommits`). The check in `onCommit` would have to then check both collections but that sounds ok to me.

Reply:
That sounds good to me :+1: I've moved the internally acquired commits to a separate map `internallyAcquiredIndexCommits`